hugolib/shortcode.go

```diff
@@ -519,49 +517,74 @@ Loop:
 
 // Replace prefixed shortcode tokens (HUGOSHORTCODE-1, HUGOSHORTCODE-2) with the real content.
 // Note: This function will rewrite the input slice.
-func replaceShortcodeTokens(source []byte, prefix string, replacements map[string]string) ([]byte, error) {
-
+func replaceShortcodeTokens(source []byte, replacements map[string]string) ([]byte, error) {
 	if len(replacements) == 0 {
 		return source, nil
 	}
 
-	sourceLen := len(source)
-	start := 0
-
-	pre := []byte("HAHA" + prefix)
+	pre := []byte("HAHA" + shortcodePlaceholderPrefix)
 	post := []byte("HBHB")
 	pStart := []byte("<p>")
 	pEnd := []byte("</p>")
 
-	k := bytes.Index(source[start:], pre)
+	preIndex := bytes.Index(source, pre)
+	if preIndex < 0 {
+		return source, nil
+	}
+
+	buf := bp.GetBuffer()
+	defer bp.PutBuffer(buf)
 
-	for k != -1 {
-		j := start + k
-		postIdx := bytes.Index(source[j:], post)
-		if postIdx < 0 {
-			// this should never happen, but let the caller decide to panic or not
-			return nil, errors.New("illegal state in content; shortcode token missing end delim")
+	lastShortcodeEnd := 0
+	for {
+		postIndex := bytes.Index(source[preIndex:], post)
+		if postIndex < 0 {
+			return nil, errors.New("illegal state in content: shortcode token missing end delim")
 		}
 
-		end := j + postIdx + 4
+		shortcodeEnd := preIndex + postIndex + len(post)
 
-		newVal := []byte(replacements[string(source[j:end])])
+		replacement := []byte(replacements[string(source[preIndex:shortcodeEnd])])
+		// If replacement also contains shortcodes.
+		if bytes.Contains(replacement, pre) {
```

Comment:
If shortcode contains another shortcode, we should process it. Without it `TestTableOfContentsInShortcodes` will not work.